### PR TITLE
CI: GitHub Actions CI job restructuring

### PR DESCRIPTION
**Pull request by rgommers**
Implements most of what is discussed in gh-24410, and documents what each job does. It removes some duplicate jobs, and adds a few notes and TODOs for where we miss coverage.

For the two files with a large number of jobs (`linux.yml` and `linux_simd.yml`) we run the same `smoke_test` which acts as a gate for the other jobs. That's an unavoidable duplication, but it takes only ~6 minutes so that should be fine. If there are only 2-3 jobs in a yml file, we don't run a smoke test. Overall, this will cut down on the number of jobs and make them easier to maintain.

### CODE REVIEW

**Comment by rgommers** [comment]
This is happy now. Should make it easier to iterate on things in parallel after this is merged. We basically have to get rid of `uses: ./.github/actions` from all these jobs, which is the part that's hitting the `setup.py` builds.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      USE_DEBUG: 1
+    steps:
+    - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+      with:
+        submodules: recursive
+        fetch-depth: 0
+    - uses: actions/setup-python@61a6322f88396a6271a6ee3565807d608ecaddd1 # v4.7.0
+      with:
+        python-version: '3.11'
+
```
Comment (charris) [inline .github/workflows/linux.yml:100]: Could get rid of the blank line at some point.

**Comment by charris** [comment]
Thanks Ralf.

**Comment by mattip** [comment]
Here are the current workflows on [github actions](https://github.com/numpy/numpy/actions) and the time they take to run (after merging this PR)
Workflow | time | Runs on merge to main
---|---|---
BLAS tests (Linux) | 9-14 min | no
CircleCI artifact redirector | 10 sec | no
CodeQL | 5-7 min | yes
Dependency Review | < 20s | no
Linux Qemu tests | 20-25 min | no
Linux tests | 19-30 min | yes
Pull Request Labeler | <20s | no
Run MyPy | 6-14 min | no
Scorecards supply-chain security | 1 min | runs only on merges
SIMD tests (Linux) | 23-26 min | no
Test Address Sanitizer Build (Linux) | 19 min | see PR #24208
Test Emscripten/Pyodide build | 10-15 min | no
Test muslinux ×86 64 | 9-10 min | no
Test on Cygwin | 34-50 min | no
Test with compiler sanitizers (Linux) | 21-32 min | no
Wheel builder | 20 min | mostly skipped, runs on a schedule
Windows tests | 14-17 min | no

In addition we have the 
- [azure workflow](https://dev.azure.com/numpy/numpy/_build?definitionId=5&_a=summary) that takes ~34-50 min
- the [cirrus builds](https://cirrus-ci.com/build/4541512705376256) (two jobs: ~12 minutes on aarch64 linux and ~3 minutes on macos-arm64) and
- the [circleCI workflow](https://app.circleci.com/pipelines/github/numpy/numpy) that takes ~23 min

**Comment by rgommers** [comment]
Thanks! That's a useful overview. Perhaps good to post it on gh-24410 too (more discoverable)? 

I think we should first get rid of all the `setup.py` files, and then see where we're at before doing anything else. We can restructure jobs, triggers, etc. after that.

**Pull request merged**